core/drivers/clk/clk-stm32-core.c

```diff
@@ -84,12 +89,24 @@ static void stm32_gate_endisable(uint16_t gate_id, bool enable)
 
 void stm32_gate_disable(uint16_t gate_id)
 {
-	stm32_gate_endisable(gate_id, false);
+	struct clk_stm32_priv *priv = clk_stm32_get_priv();
+	uint8_t *gate_cpt = priv->gate_cpt;
+
+	assert(gate_cpt[gate_id] > 0);
+	if (gate_cpt[gate_id] == 1)
+		stm32_gate_endisable(gate_id, false);
+	gate_cpt[gate_id]--;
```

Comment:
We need some concurrent access protection here, around the counter update and call of `stm32_gate_endisable()`. `cpu_spin_lock_xsave()`/`cpu_spin_lock_xrestore()` with a common lock reference should do the work.